humacli/humacli.go

```diff
@@ -0,0 +1,308 @@
+//			fmt.Println("Hello " + opts.Name)
+//		}),
+//	})
+func WithOptions[Options any](f func(cmd *cobra.Command, args []string, options *Options)) func(*cobra.Command, []string) {
+	return func(cmd *cobra.Command, s []string) {
+		var options *Options = cmd.Context().Value(optionsKey).(*Options)
+		f(cmd, s, options)
+	}
+}
+
+type option struct {
+	name string
+	typ  reflect.Type
+	path []int
+}
+
+type cli[Options any] struct {
+	root     *cobra.Command
+	optInfo  []option
+	onParsed func(Hooks, *Options)
+	start    func()
+	stop     func()
+}
+
+func (c *cli[Options]) Run() {
+	var o Options
+
+	existing := c.root.PersistentPreRun
+	c.root.PersistentPreRun = func(cmd *cobra.Command, args []string) {
+		// Load config from args/env/files
+		v := reflect.ValueOf(&o).Elem()
+		flags := c.root.PersistentFlags()
+		for _, opt := range c.optInfo {
+			f := v
+			for _, i := range opt.path {
+				f = f.Field(i)
+			}
+			var fv reflect.Value
+			switch deref(opt.typ).Kind() {
+			case reflect.String:
+				s, _ := flags.GetString(opt.name)
+				fv = reflect.ValueOf(s)
+			case reflect.Int, reflect.Int64:
+				var i any
+				if opt.typ == durationType {
+					i, _ = flags.GetDuration(opt.name)
+				} else {
+					i, _ = flags.GetInt64(opt.name)
+				}
+				fv = reflect.ValueOf(i).Convert(deref(opt.typ))
+			case reflect.Bool:
+				b, _ := flags.GetBool(opt.name)
+				fv = reflect.ValueOf(b)
+			}
+
+			if opt.typ.Kind() == reflect.Ptr {
+				ptr := reflect.New(fv.Type())
+				ptr.Elem().Set(fv)
+				fv = ptr
+			}
+
+			f.Set(fv)
+		}
+
+		// Run the parsed callback.
+		c.onParsed(c, &o)
+
+		if existing != nil {
+			existing(cmd, args)
+		}
+
+		// Set options in context, so custom commands can access it.
+		cmd.SetContext(context.WithValue(cmd.Context(), optionsKey, &o))
+	}
+
+	// Run the command!
+	c.root.Execute()
+}
+
+func (c *cli[O]) Root() *cobra.Command {
+	return c.root
+}
+
+func (c *cli[O]) OnStart(fn func()) {
+	c.start = fn
+}
+
+func (c *cli[O]) OnStop(fn func()) {
+	c.stop = fn
+}
+
+func (c *cli[O]) setupOptions(t reflect.Type, path []int) {
+	var err error
+	flags := c.root.PersistentFlags()
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+
+		if !field.IsExported() {
+			// This isn't a public field, so we cannot use reflect.Value.Set with
+			// it. This is usually a struct field with a lowercase name.
+			fmt.Println("warning: ignoring unexported options field", field.Name)
+			continue
+		}
+
+		currentPath := append([]int{}, path...)
+		currentPath = append(currentPath, i)
+
+		fieldType := deref(field.Type)
+		if field.Anonymous {
+			// Embedded struct. This enables composition from e.g. company defaults.
+			c.setupOptions(fieldType, currentPath)
+			continue
+		}
+
+		name := field.Tag.Get("name")
+		if name == "" {
+			name = casing.Kebab(field.Name)
+		}
+
+		envName := "SERVICE_" + casing.Snake(name, strings.ToUpper)
+		defaultValue := field.Tag.Get("default")
+		if v := os.Getenv(envName); v != "" {
```

Comment:
I think `os.LookupEnv` is better choice. Because with it we can override default value with empty string if needed.
```go
if v, ok := os.LookupEnv(envName); ok {
```